### PR TITLE
GH-107674: Avoid allocating boxed ints for `sys.settrace` line events

### PR DESCRIPTION
**Pull request by markshannon**
The [benchmarks](https://github.com/faster-cpython/benchmarking-public/tree/main/results/bm-20230805-3.13.0a0-83ffca9) are noisy but show a 6% performance increase for coverage.

* Issue: gh-107674

### CODE REVIEW

File: Python/instrumentation.c
Comment (ericsnowcurrently) [inline Python/instrumentation.c:1186]: Would it make sense to short-circuit here, to avoid create the int object?

```suggestion
    }
    if (tools == 0) {
        goto done;
    }
    PyObject *line_obj = PyLong_FromLong(line);
```

File: Python/instrumentation.c
Comment (ericsnowcurrently) [inline Python/instrumentation.c]: It would be helpful to have a comment that identifies where to find the original code that you copy-and-pasted here.

If I've understood correctly, that would be `call_one_instrument()`, along with the registered instrument (wherever that code is).

**Review by ericsnowcurrently — APPROVED** [review]
Aside from a couple minor comments, LGTM.

(I'm approving, assuming you'll address the comments before merging.)

File: Python/instrumentation.c
Comment (markshannon) [reply Python/instrumentation.c:1186]: Yes it would. Done

File: Python/instrumentation.c
Comment (markshannon) [reply Python/instrumentation.c]: The code isn't directly copied from anywhere, but it does have to do the same preparation and clean up as `call_one_instrument()`. I've added a couple of comments.

**Pull request merged**
